src/publish-registry.ts

```diff
@@ -30,46 +30,52 @@ export default async function publishRegistry(dt: FS, allPackages: AllPackages,
+    assert.strictEqual(npmVersion.major, 0);
+    assert.strictEqual(npmVersion.minor, 1);
 
     // Don't include not-needed packages in the registry.
     const registryJsonData = await withNpmCache(client, cachedClient => generateRegistry(allPackages.allLatestTypings(), cachedClient));
     const registry = JSON.stringify(registryJsonData);
     const newContentHash = computeHash(registry);
+    const newVersion = `0.1.${npmVersion.patch + 1}`;
 
-    assert.strictEqual(oldVersion.major, 0);
-    assert.strictEqual(oldVersion.minor, 1);
-    const newVersion = `0.1.${oldVersion.patch + 1}`;
-    const packageJson = generatePackageJson(newVersion, newContentHash);
-    await generate(registry, packageJson);
-
-    const publishClient = () => NpmPublishClient.create({ defaultTag: "next" });
-    if (!highestSemverVersion.equals(oldVersion)) {
-        // There was an error in the last publish and types-registry wasn't validated.
-        // This may have just been due to a timeout, so test if types-registry@next is a subset of the one we're about to publish.
-        // If so, we should just update it to "latest" now.
-        log("Old version of types-registry was never tagged latest, so updating");
-        await validateIsSubset(await readNotNeededPackages(dt), log);
-        await (await publishClient()).tag(packageName, highestSemverVersion.versionString, "latest");
-    } else if (oldContentHash !== newContentHash && isAWeekAfter(lastModified)) {
-        log("New packages have been added, so publishing a new registry.");
-        await publish(await publishClient(), packageJson, newVersion, dry, log);
-    } else {
-        const reason = oldContentHash === newContentHash ? "Was modified less than a week ago" : "No new packages published";
-        log(`${reason}, so no need to publish new registry.`);
-        // Just making sure...
-        await validate(log);
+    await publishToRegistry("github");
+    await publishToRegistry("npm");
+    await writeLog("publish-registry.md", logResult());
+
+    async function publishToRegistry(registryName: "github" | "npm") {
+        const packageName = registryName === "github" ? "@definitelytyped/" + typesRegistry : typesRegistry;
+        const packageJson = generatePackageJson(packageName, registryName, newVersion, newContentHash);
+        await generate(registry, packageJson);
+
+        const publishClient = () => NpmPublishClient.create({ defaultTag: "next" }, registryName);
+        if (!highestSemverVersion.equals(npmVersion)) {
+            // There was an error in the last publish and types-registry wasn't validated.
+            // This may have just been due to a timeout, so test if types-registry@next is a subset of the one we're about to publish.
+            // If so, we should just update it to "latest" now.
+            log("Old version of types-registry was never tagged latest, so updating");
+            await validateIsSubset(await readNotNeededPackages(dt), log);

```

Comment:
Does this throw? An error here (or anywhere in this function) publishing to GitHub will stop us from attempting a publish to npm. Not sure whether that’s intentional or not. If the goal is to keep them in sync, that seems good. If the goal is to try publishing to GitHub without effect on current npm infrastructure, that seems bad.

Reply:
Yes! It does. The intention is to avoid publishing whenever anything goes wrong even if we're not quite sure why, since an old types-registry is way less bad than a new wrong one.

I don't think it's a problem here, but I'll need to inspect the rest of the code.
> (I inspected a bunch of code and concluded that there WERE a few problems that got complicated to fix)

I think the safest thing here is to make sure, as long as we're publishing to two places, that github never gets ahead of npm. It's fine if github misses some updates compared to npm, but not the other way around since we always check against npm. That also means, unfortunately, that github needs to update first so that `validateIsSubset` checks against the current version of npm, not current+1.

I think the right fix is to surround `await publishToRegistry('github')` with `try { ... } catch { }`, so that if github fails, npm tries as well.

Reply:
That was my hunch (that npm should still try to publish even if GitHub fails), but wanted to double check.